### ISSUE TITLE
start + end outside length of audio

### ISSUE THREAD

**Issue by doublex** [Jeronymous/whisper-timestamped#6]
This 15s audio: [gaenswein15.zip](https://github.com/Jeronymous/whisper-timestamped/files/10433614/gaenswein15.zip)

Command line:
````
python3 whisper_timestamped/transcribe.py ~/gaenswein15.wav --model large-v2 --language de
````

Timestamps:
````json
    {
      "id": 2,
      "seek": 1300,
      "start": 27.16,
      "end": 27.86,
      "text": " Das hat er als emeritus Ritus gewünscht.",
 ````

**Comment by Jeronymous** [comment]
Thank you @doublex  for reporting and giving a simple way to reproduce (I really appreciate that).
I'll investigate this after having fixed issue #4

**Comment by Jeronymous** [comment]
This seems to be mainly due to the use of "large-v2", which predicts some nonsense timestamps.
```
$ whisper gaenswein15.mp3 --model large-v2 --language de

[00:00.000 --> 00:09.000]  Die Wiederzulassung des Messbuchs von 1962 als Missale für die außerordentliche Form des römischen Ritus
[00:09.000 --> 00:13.000]  ist dann nicht so weitergegangen, wie sich Papst Benedikt das gewünscht hatte.
[00:13.000 --> 00:31.000]  Das hatte er als emeritus Ritual.
```
(The last "00:31.000" is outside the boundary)

I had other issues with the upgrade from large-v1 to large-v2, and I heard other complaints about v2 giving worse results than v1.
So first, I advise to use model large-v1 instead (if not medium).

Concerning the bug, I think that the main could/should be fixed in [openai-whisper](https://github.com/openai/whisper) itself (unfortunately, there does not seem to be any "Issue" section on this repo...).
The decoding process should prevent from predicting timestamps outside of signal boundaries (there is already a mechanism to exclude some predictions at each decoding step, some additional constraints should be added to take into account maximal duration on the padded input signal).

I'm watching what I can do in `whisper-timestamped`.
I am not sure that the general problem can be fixed without touching whisper (maybe I'll end up forking whisper in the end, 'cause I am already hacking and retro-engineering it a lot...).
But in the particular case of this issue, it's only the last end timestamp that is wrong, so maybe I can do something...
(if the start timestamp of the last segment would be outside the boundary, it would be much worse...)

**Comment by Jeronymous** [comment]
@doublex Do you allow me to add an mp3 version of the audio you sent here in the repo (for testing purpose in the ["test/data" folder](https://github.com/Jeronymous/whisper-timestamped/tree/master/tests/data))?

**Comment by Jeronymous** [comment]
OK this is much better now.
I pushed an improvement that solves the issue on your audio.
(I'm still interested to know if I include that audio in the testsuite)

**Comment by doublex** [comment]
Thanks a lot! Great code!
The interview was public - but unfortunately the recording is not mine.
I'm not a lawyer.

**Comment by Tangyiming205069** [comment]
Facing the same issue, I cannot upload the zip audio since it is too big. Using large-v2, I have 30 audio files and face some out of bound issue on some file and some even (the start timestamp of the last segment would be outside the boundary), as you mentioned. Then, I used large-v1 to reprocess all files, but same thing happen on some other files (the file look good when using v2). I set vad = True and recompute_all_timestamps as default (False)
<img width="383" alt="Screenshot 2023-05-16 at 15 39 02" src="https://github.com/linto-ai/whisper-timestamped/assets/127170220/623a9c0a-cbab-4c92-9395-d8d4467bcdbe">
<img width="331" alt="Screenshot 2023-05-16 at 15 39 13" src="https://github.com/linto-ai/whisper-timestamped/assets/127170220/a1d08bcd-01ea-4e0e-9a8b-b7b7f2283df7">
The pics show the result from whisper and the true lenght of the audio.

**Comment by Tangyiming205069** [comment]
1. No such warnings. 
2. I have 30 files. Without VAD, the 03 file does not have the issue any more but occur at other files. (with VAD or using large-v2, out of bound happens on different files) The first is with VAD, the second is without. simple means there exists a word ending out of bound, severe means the start of the word out of bound. The second one cannot be screenshot in one pics. Basically, there are 6 files out of bound and the others are good.
<img width="481" alt="Screenshot 2023-05-17 at 21 27 47" src="https://github.com/linto-ai/whisper-timestamped/assets/127170220/f4a67811-cacd-4d73-8506-3ca0e223b1f2">
<img width="406" alt="Screenshot 2023-05-17 at 21 29 55" src="https://github.com/linto-ai/whisper-timestamped/assets/127170220/83595507-735b-42b3-9d15-accb5a2c45ea">


3. I do have a file recored by VAD. It records who is talking at each frame. When transfer seconds to frames, there exists a word at range of no one speaking. It happens more frequently, so I develop a way to assign word to speaker when no speaking.

**Comment by ethanzrd** [comment]
Hey!

Regarding this issue:

I don't think I can reproduce it, sometimes it happens, sometimes it doesn't. I have an application that constantly transcribes 2-second long batches and the start + end times really overshoot, using the small model. Here's an example:

```python
{
    "text": " I'm going to see what's up. No it's also...",
    "segments": [
        {
            "id": 0,
            "seek": 0,
            "start": 19.7,
            "end": 29.14,
            "text": " I'm going to see what's up. No it's also...",
            "tokens": [
                50363,
                314,
                1101,
                1016,
                284,
                766,
                644,
                338,
                510,
                13,
                1400,
                340,
                338,
                635,
                986,
            ],
            "temperature": 0.0,
            "avg_logprob": -0.2860553562641144,
            "compression_ratio": 0.8958333333333334,
            "no_speech_prob": 0.07040286064147949,
            "confidence": 0.78,
            "words": [
                {"text": "I'm", "start": 19.7, "end": 24.82, "confidence": 0.588},
                {"text": "going", "start": 24.82, "end": 24.84, "confidence": 0.766},
                {"text": "to", "start": 24.84, "end": 24.86, "confidence": 0.994},
                {"text": "see", "start": 24.86, "end": 24.88, "confidence": 0.953},
                {"text": "what's", "start": 24.88, "end": 27.68, "confidence": 0.971},
                {"text": "up.", "start": 27.68, "end": 27.7, "confidence": 0.997},
                {"text": "No", "start": 27.7, "end": 27.72, "confidence": 0.624},
                {"text": "it's", "start": 27.72, "end": 28.94, "confidence": 0.922},
                {"text": "also...", "start": 28.94, "end": 29.14, "confidence": 0.405},
            ],
        }
    ],
    "language": "en",
}
```

However, when transcribed separately (not following previous batches), it still overshoots, but not as hard:
```python
{
    "text": " to see what's up. No, it's also a",
    "segments": [
        {
            "id": 0,
            "seek": 0,
            "start": 0.0,
            "end": 0.74,
            "text": " to see what's up.",
            "tokens": [50363, 284, 766, 644, 338, 510, 13, 50413],
            "temperature": 0.0,
            "avg_logprob": -0.9967834299260919,
            "compression_ratio": 0.9523809523809523,
            "no_speech_prob": 0.39323943853378296,
            "confidence": 0.302,
            "words": [
                {"text": "to", "start": 0.0, "end": 0.22, "confidence": 0.028},
                {"text": "see", "start": 0.22, "end": 0.32, "confidence": 0.171},
                {"text": "what's", "start": 0.32, "end": 0.52, "confidence": 0.738},
                {"text": "up.", "start": 0.52, "end": 0.74, "confidence": 0.954},
            ],
        },
        {
            "id": 1,
            "seek": 0,
            "start": 1.38,
            "end": 2.24,
            "text": " No, it's also a",
            "tokens": [50413, 1400, 11, 340, 338, 635, 257, 50463],
            "temperature": 0.0,
            "avg_logprob": -0.9967834299260919,
            "compression_ratio": 0.9523809523809523,
            "no_speech_prob": 0.39323943853378296,
            "confidence": 0.409,
            "words": [
                {"text": "No,", "start": 1.38, "end": 1.58, "confidence": 0.253},
                {"text": "it's", "start": 1.66, "end": 1.82, "confidence": 0.892},
                {"text": "also", "start": 1.82, "end": 2.04, "confidence": 0.292},
                {"text": "a", "start": 2.04, "end": 2.24, "confidence": 0.194},
            ],
        },
        {
            "id": 2,
            "seek": 0,
            "start": 2.0,
            "end": 3.5,
            "text": " It's a",
            "tokens": [632, 338, 257],
            "temperature": 0.0,
            "avg_logprob": -0.9967834299260919,
            "compression_ratio": 0.9523809523809523,
            "no_speech_prob": 0.39323943853378296,
            "words": [
                {"text": "It's", "start": 2.84, "end": 3.48, "confidence": 0.215},
                {"text": "a", "start": 3.48, "end": 3.5, "confidence": 0.464},
            ],
            "confidence": 0.278,
        },
    ],
    "language": "en",
}

```

[Audio file used (in the first example, it was transcribed using its waveform)](https://voca.ro/14TFRzkONtrJ)

Happens pretty often with other audio files. Would love to get your opinion :)

**Comment by chenkigba** [comment]
> Thank you for reporting 感谢你报告[@Tangyiming205069](https://github.com/Tangyiming205069) It's a pity that this can occur.很遗憾这个问题会发生。
> 
> Some questions that can help to investigate:一些有助于调查的问题：
> 
> * Do you know if there were some warnings `Got start time outside of audio boundary` when running the script?运行脚本时你是否知道是否有警告 `Got start time outside of audio boundary` ？
> * Do you know if you have the same problem without VAD?没有 VAD 你是否知道会出现同样的问题？
> * Do you know if it occurs on parts where Whisper "hallucinates" (i.e. generates stuff that corresponds to nothing in the input audio)?你是否知道它是否发生在 Whisper 产生幻觉的部分（即生成的内容与输入音频中的任何内容都不对应）？

Hi, also meet this problem in my case.

Here is some clue to help understanding why this happenning sometimes.

Situation 1: No Warning like `Got start time outside of audio boundary` 

When I use librosa to remove silence segments, there never been Warnings like outside of audio boundary.

Codes like: 

```python
def remove_silence_librosa_trim(input_path, folder="silence_removed", top_db=20):
    y, sr = librosa.load(input_path, sr=None)
    y_trimmed, _ = librosa.effects.trim(y, top_db=top_db)
    output_path = _get_output_path_in_folder(input_path, folder)
    sf.write(output_path, y_trimmed, sr)
    return y_trimmed, output_path.replace('\\', '/')
```

Situation 2: Warning like `Got start time outside of audio boundary` shows up many times

When I didn't preprocess audios with librosa.

![Image](https://github.com/user-attachments/assets/ecff79cb-3013-4103-a399-9809760c3f99)

So, I assume it is related with SILENCE.

We already know that whisper sometimes attempts to let first char's start_time to 00:00, I think these two problem is related.

**Comment by chenkigba** [comment]
The following is AI analysis of why we get `Got start time outside of audio boundary`.

TLTR: It is related with Whisper hallucination in silent part. Use vad is a solution.

Hello! The warning message 'Got start time outside of audio boundary' appears in the performac_word_alignment function of the whisper_timestamped library.
By analyzing the code you provided, we can draw the following conclusions:
The triggering condition for this warning is start_token>=max_dauration. To understand this condition, we need to understand the meanings of the variables start_token and max_dauration:
Start_token: This is the start time of a speech segment predicted by the Whisper model. Internally, this time is converted into a unit called 'token time'. It comes from the first special token generated by the model for each recognized speech segment, which itself represents a timestamp (e.g.<| 10.24 |>).
Max_dauration: This is the actual effective audio duration in the current audio block being processed, also measured in 'token time'. Whisper timestamped divides audio into 30 second blocks when processing it. If an audio block (especially the end of a file) is less than 30 seconds, it will be muted and filled to reach a length of 30 seconds. Max_dauration is the process of analyzing audio features (MFCC) to determine the starting position of the mute fill and ultimately the end position of the actual audio.
Therefore, the fundamental reason for this warning is:
The start time of a speech segment predicted by the Whisper model falls within the mute fill area of the current audio block. In other words, the model "hallucinates" and generates text with timestamps in places without real audio signals.
A typical scenario is:
When processing the last segment of an audio file, if the total duration of the file is not an integer multiple of 30 seconds, the last audio block will be shorter than 30 seconds and filled with mute. At this point, if the Whisper model fails to accurately stop at the end of the audio and continues to generate text and timestamps in the silent area, then this timestamp will exceed the boundary of the real audio (start_token>=max_dauration), triggering this warning.
Simply put, this warning indicates that the model may have generated inaccurate timestamps located in the quiet zone at the end of the audio.
